# OpenAI cartpole basics

https://www.gymlibrary.dev/environments/classic_control/cart_pole/

https://www.gymlibrary.dev/environments/classic_control/

In [15]:
!pip install gym==0.17.3

     ---------------------------------------- 0.0/1.6 MB ? eta -:--:--
     ----------- ---------------------------- 0.5/1.6 MB 9.4 MB/s eta 0:00:01
     ------------------------------ --------- 1.2/1.6 MB 15.4 MB/s eta 0:00:01
     ---------------------------------------- 1.6/1.6 MB 14.3 MB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
     ---------------------------------------- 0.0/1.0 MB ? eta -:--:--
     ---------------------- ----------------- 0.6/1.0 MB 12.2 MB/s eta 0:00:01
     ---------------------------------------- 1.0/1.0 MB 13.0 MB/s eta 0:00:00
     ---------------------------------------- 0.0/840.9 kB ? eta -:--:--
     ------------------------------------  839.7/840.9 kB 26.8 MB/s eta 0:00:01
     ------------------------------------- 840.9/840.9 kB 17.7 MB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Created wheel for gym: fi

The action is a ndarray with shape (1,) which can take values {0, 1} indicating the direction of the fixed force the cart is pushed with.

Num Action (apply a force)

0  Push cart to the left

1  Push cart to the right

rewards
Since the goal is to keep the pole upright for as long as possible, a reward of +1 for every step taken, including the termination step, is allotted. The threshold for rewards is 475 for v1.

episode is max of 10s, if can remain upright thru the entire time then episode done

Episode End
The episode ends if any one of the following occurs:

Termination: Pole Angle is greater than ±12°

Termination: Cart Position is greater than ±2.4 (center of the cart reaches the edge of the display)

Truncation: Episode length is greater than 500 (200 for v0)
    
Starting State
All observations are assigned a uniformly random value in (-0.05, 0.05)

Arguments
gym.make('CartPole-v1')

In [45]:
import gym

env = gym.make('CartPole-v1',render_mode='human')
env.reset()

# Display the environment - will pop up in a window in a separate env
while True:
    action=env.action_space.sample()
    env.step(action)
    env.render()

Unexpected exception formatting exception. Falling back to standard exception


Traceback (most recent call last):
  File "C:\Users\Palaash.HPZ\anaconda3\envs\pytorch_env\lib\site-packages\IPython\core\interactiveshell.py", line 3508, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "C:\Users\Palaash.HPZ\AppData\Local\Temp\ipykernel_6028\3641482266.py", line 9, in <module>
    env.step(action)
  File "C:\Users\Palaash.HPZ\anaconda3\envs\pytorch_env\lib\site-packages\gym\wrappers\time_limit.py", line 50, in step
  File "C:\Users\Palaash.HPZ\anaconda3\envs\pytorch_env\lib\site-packages\gym\wrappers\order_enforcing.py", line 37, in step
  File "C:\Users\Palaash.HPZ\anaconda3\envs\pytorch_env\lib\site-packages\gym\wrappers\env_checker.py", line 39, in step
  File "C:\Users\Palaash.HPZ\anaconda3\envs\pytorch_env\lib\site-packages\gym\envs\classic_control\cartpole.py", line 187, in step
    pole.add_attr(self.carttrans)
  File "C:\Users\Palaash.HPZ\anaconda3\envs\pytorch_env\lib\site-packages\gym\envs\classic_control\cartpole.py", line 298, in re

In [46]:
env.close()

# Q-learning exploration
on OpenAI Frozen-Lake

DeepLizard q learning algo tutorial
https://www.youtube.com/watch?v=HGeI30uATws&list=PLZbbT5o_s2xoWNVdDudn51XM8lOuZ_Njv&index=9&ab_channel=deeplizard

In [1]:
!pip install gym

Frozen Lake
I've grabbed the description of the game directly from Gym's website. Let's read through it together.

Winter is here. You and your friends were tossing around a frisbee at the park when you made a wild throw that left the frisbee out in the middle of the lake. The water is mostly frozen, but there are a few holes where the ice has melted. If you step into one of those holes, you'll fall into the freezing water. At this time, there's an international frisbee shortage, so it's absolutely imperative that you navigate across the lake and retrieve the disc. However, the ice is slippery, so you won't always move in the direction you intend. The surface is described using a grid like the following:

SFFF
FHFH
FFFH
HFFG
This grid is our environment where S is the agent's starting point, and it's safe. F represents the frozen surface and is also safe. H represents a hole, and if our agent steps in a hole in the middle of a frozen lake, well, that's not good. Finally, G represents the goal, which is the space on the grid where the prized frisbee is located.

The agent can navigate left, right, up, and down, and the episode ends when the agent reaches the goal or falls in a hole. It receives a reward of one if it reaches the goal, and zero otherwise.

State	Description	Reward
S	Agent's starting point - safe	0
F	Frozen surface - safe	0
H	Hole - game over	0
G	Goal - game over	1

In [2]:
import numpy as np
import gym
import random
import time
from IPython.display import clear_output

In [3]:
env=gym.make('FrozenLake-v0')
env.reset()
env.render()


SFFF
FHFH
FFFH
HFFG


In [7]:
action_space_size=env.action_space.n
state_space_size=env.observation_space.n
action_space_size,state_space_size

(4, 16)

In [2]:
q_table=np.zeros((state_space_size,action_space_size))
q_table #cols are actions, rows are state spaces

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [14]:
n_episodes=10000
max_steps=100 #per episode

lr=0.1
gamma=0.99 #discount rate

epsilon=1 #exploration rate initialised to 1
max_epsilon=1
min_epsilon=0.01
epsilon_decay_rate=0.001

In [26]:
rewards_all_episodes=[]

env=gym.make('FrozenLake-v0')
env.reset()

for episode in range(n_episodes):
    state=env.reset()
    
    done=False
    rewards_current=0
    
    for step in range(max_steps):
        
        #exploration vs exploitation trade off
        epsilon_random_num=random.uniform(0,1) #generate random number and compare to epsilon to see whether to explore or exploit
        if epsilon_random_num>epsilon:
            action=np.argmax(q_table[state,:]) #exploit - get the action with the best q value for that given state
        else:
            action=env.action_space.sample() #explore - get a random action to explore the env
        
        new_state,reward,done,info=env.step(action)
        
        #update q values
        q_table[state,action]=q_table[state,action] * (1-lr) + lr*(reward+gamma*np.max(q_table[new_state,:])) #follow the formula
        
        state=new_state
        rewards_current+=reward
        
        if done==True:
            break
            
    #exploration rate decay
    epsilon=min_epsilon+(max_epsilon-min_epsilon)*np.exp(-epsilon*episode)

    rewards_all_episodes.append(rewards_current)
    
    
# Calculate and print the average reward per thousand episodes
rewards_per_thousand_episodes = np.split(np.array(rewards_all_episodes),n_episodes/1000)
count = 1000

print("********Average reward per thousand episodes********\n")
for r in rewards_per_thousand_episodes:
    print(count, ": ", str(sum(r/1000)))
    count += 1000
    
# Print updated Q-table
print("\n\n********Q-table********\n")
print(q_table)

********Average reward per thousand episodes********

1000 :  0.1300000000000001
2000 :  0.4110000000000003
3000 :  0.4250000000000003
4000 :  0.5830000000000004
5000 :  0.6960000000000005
6000 :  0.6830000000000005
7000 :  0.6830000000000005
8000 :  0.6500000000000005
9000 :  0.6930000000000005
10000 :  0.7100000000000005


********Q-table********

[[0.55286845 0.51268178 0.49434934 0.50160874]
 [0.08360022 0.11746611 0.15124224 0.49224999]
 [0.39742406 0.19640409 0.17165541 0.20788522]
 [0.11904098 0.         0.         0.        ]
 [0.56502277 0.45482487 0.35033856 0.36754067]
 [0.         0.         0.         0.        ]
 [0.34978046 0.12790535 0.16617634 0.09619136]
 [0.         0.         0.         0.        ]
 [0.33754695 0.44879802 0.40354035 0.60304211]
 [0.45907248 0.67378309 0.54589134 0.43403797]
 [0.64578107 0.24956514 0.34572227 0.39508101]
 [0.         0.         0.         0.        ]
 [0.         0.         0.         0.        ]
 [0.40154229 0.56086349 0.80261477 0.

ave reward of 0.1300000000000001 means the agent hits the target 13% of the time (13%chance of hitting target) 

In [37]:
# try put the code into functions

def train_one_episode(max_steps,state,epsilon,q_table):
    done=False
    rewards_current_episode=0
    for step in range(max_steps):
        
        #exploration vs exploitation trade off
        epsilon_random_num=random.uniform(0,1) #generate random number and compare to epsilon to see whether to explore or exploit
        if epsilon_random_num>epsilon:
            action=np.argmax(q_table[state,:]) #exploit - get the action with the best q value for that given state
        else:
            action=env.action_space.sample() #explore - get a random action to explore the env
        
        new_state,reward,done,info=env.step(action)
        
        #update q values - follow the formula
        q_table[state,action]=q_table[state,action] * (1-lr) + lr*(reward+gamma*np.max(q_table[new_state,:])) 
        
        state=new_state
        rewards_current_episode+=reward
        
        if done:
            break
            
        return state,rewards_current_episode,q_table
    
# training loop

n_episodes=10000
max_steps=100 #per episode

lr=0.1
gamma=0.99 #discount rate

epsilon=1 #exploration rate initialised to 1
max_epsilon=1
min_epsilon=0.01
epsilon_decay_rate=0.001

env=gym.make('FrozenLake-v0')
env.reset()
rewards_all_episodes=[]

print_count=1000
for episode in range(n_episodes):
    state=env.reset()
    #train 1 episode - update current state, entire current episode rewards, q_table
    state,rewards_current_episode,q_table=train_one_episode(max_steps,state,epsilon,q_table) 
    rewards_all_episodes.append(rewards_current_episode)
    
    #exploration rate decay
    epsilon=min_epsilon+(max_epsilon-min_epsilon)*np.exp(-epsilon*episode)
    
    #print results
    if (episode+1)%print_count==0:
        batch_rewards=rewards_all_episodes[print_count-1000:print_count]
        ave_rewards_last_batch=sum(batch_rewards)/1000
        print(f'Average rewards for Episode {print_count-1000}-{print_count}: {ave_rewards_last_batch}')
        print_count+=1000

In [1]:
import numpy as np
import gym
import random
import time
from IPython.display import clear_output

env=gym.make('FrozenLake-v0')
env.reset()

action_space_size=env.action_space.n
state_space_size=env.observation_space.n

q_table=np.zeros((state_space_size,action_space_size)) #cols are actions, rows are state spaces


n_episodes=20000
max_steps=100 #per episode

lr=0.1
gamma=0.99 #discount rate

epsilon=1 #exploration rate initialised to 1
max_epsilon=1
min_epsilon=0.01
epsilon_decay_rate=0.001

rewards_all_episodes=[]
start_time=time.time()
for episode in range(n_episodes):
    state=env.reset()
    
    done=False
    rewards_current=0
    
    for step in range(max_steps):
        
        #exploration vs exploitation trade off
        epsilon_random_num=random.uniform(0,1) #generate random number and compare to epsilon to see whether to explore or exploit
        if epsilon_random_num>epsilon:
            action=np.argmax(q_table[state,:]) #exploit - get the action with the best q value for that given state
        else:
            action=env.action_space.sample() #explore - get a random action to explore the env
        
        new_state,reward,done,info=env.step(action)
        
        #update q values
        q_table[state,action]=q_table[state,action] * (1-lr) + lr*(reward+gamma*np.max(q_table[new_state,:])) #follow the formula
        
        state=new_state
        rewards_current+=reward
        
        if done==True:
            break
            
    #exploration rate decay
    epsilon=min_epsilon+(max_epsilon-min_epsilon)*np.exp(-epsilon*episode)

    rewards_all_episodes.append(rewards_current)
    
    if (episode+1)%1000==0:
        ave_reward_last_thousand_ep=sum(rewards_all_episodes[-1000:])/1000
        print(f'Average reward for Episode {episode+1-1000}-{episode+1}: {ave_reward_last_thousand_ep}, Time elapsed: {(time.time()-start_time):.2f}s')
    
    
# # Calculate and print the average reward per thousand episodes
# rewards_per_thousand_episodes = np.split(np.array(rewards_all_episodes),n_episodes/1000)
# count = 1000

# print("********Average reward per thousand episodes********\n")
# for r in rewards_per_thousand_episodes:
#     print(count, ": ", str(sum(r/1000)))
#     count += 1000
    
# Print updated Q-table
print("\n\n********Q-table********\n")
print(q_table)

Average reward for Episode 0-1000: 0.086, Time elapsed: 0.26s
Average reward for Episode 1000-2000: 0.178, Time elapsed: 0.66s
Average reward for Episode 2000-3000: 0.399, Time elapsed: 1.52s
Average reward for Episode 3000-4000: 0.43, Time elapsed: 2.22s
Average reward for Episode 4000-5000: 0.52, Time elapsed: 3.51s
Average reward for Episode 5000-6000: 0.662, Time elapsed: 4.40s
Average reward for Episode 6000-7000: 0.633, Time elapsed: 5.71s
Average reward for Episode 7000-8000: 0.681, Time elapsed: 7.22s
Average reward for Episode 8000-9000: 0.667, Time elapsed: 8.65s
Average reward for Episode 9000-10000: 0.693, Time elapsed: 10.13s
Average reward for Episode 10000-11000: 0.661, Time elapsed: 11.61s
Average reward for Episode 11000-12000: 0.681, Time elapsed: 13.18s
Average reward for Episode 12000-13000: 0.637, Time elapsed: 14.64s
Average reward for Episode 13000-14000: 0.665, Time elapsed: 16.21s
Average reward for Episode 14000-15000: 0.668, Time elapsed: 17.64s
Average rewar

# Q-learning - full code

In [2]:
# full code 

import numpy as np
import gym
import random
import time
from IPython.display import clear_output

env=gym.make('FrozenLake-v0')
env.reset()

action_space_size=env.action_space.n
state_space_size=env.observation_space.n

q_table=np.zeros((state_space_size,action_space_size)) #cols are actions, rows are state spaces


n_episodes=50000
max_steps=100 #per episode

lr=0.1
gamma=0.99 #discount rate

epsilon=1 #exploration rate initialised to 1
max_epsilon=1
min_epsilon=0.01
epsilon_decay_rate=0.001

rewards_all_episodes=[]
start_time=time.time()
for episode in range(n_episodes):
    state=env.reset()
    
    done=False
    rewards_current=0
    
    for step in range(max_steps):
        
        #exploration vs exploitation trade off
        epsilon_random_num=random.uniform(0,1) #generate random number and compare to epsilon to see whether to explore or exploit
        if epsilon_random_num>epsilon:
            action=np.argmax(q_table[state,:]) #exploit - get the action with the best q value for that given state
        else:
            action=env.action_space.sample() #explore - get a random action to explore the env
        
        new_state,reward,done,info=env.step(action)
        
        #update q values
        q_table[state,action]=q_table[state,action] * (1-lr) + lr*(reward+gamma*np.max(q_table[new_state,:])) #follow the formula
        
        state=new_state
        rewards_current+=reward
        
        if done==True:
            break
            
    #exploration rate decay
    epsilon=min_epsilon+(max_epsilon-min_epsilon)*np.exp(-epsilon*episode)

    rewards_all_episodes.append(rewards_current)
    
    if (episode+1)%5000==0:
        ave_reward_last_thousand_ep=sum(rewards_all_episodes[-5000:])/5000
        print(f'Average reward for Episode {episode+1-5000}-{episode+1}: {ave_reward_last_thousand_ep}, Time elapsed: {(time.time()-start_time):.2f}s')
    
    
# Print updated Q-table
print("\n\n********Q-table********\n")
print(q_table)

Average reward for Episode 0-5000: 0.3964, Time elapsed: 5.67s
Average reward for Episode 5000-10000: 0.6624, Time elapsed: 12.40s
Average reward for Episode 10000-15000: 0.6504, Time elapsed: 19.32s
Average reward for Episode 15000-20000: 0.6736, Time elapsed: 26.45s
Average reward for Episode 20000-25000: 0.6722, Time elapsed: 33.68s
Average reward for Episode 25000-30000: 0.6828, Time elapsed: 40.55s
Average reward for Episode 30000-35000: 0.6646, Time elapsed: 47.56s
Average reward for Episode 35000-40000: 0.6796, Time elapsed: 54.72s
Average reward for Episode 40000-45000: 0.6744, Time elapsed: 61.73s
Average reward for Episode 45000-50000: 0.683, Time elapsed: 68.86s


********Q-table********

[[0.49242512 0.48833038 0.47135847 0.47490835]
 [0.2654378  0.30636939 0.33686358 0.44115769]
 [0.35838883 0.2564736  0.24168812 0.22036891]
 [0.12514671 0.         0.         0.        ]
 [0.52177627 0.4403537  0.44295825 0.42620948]
 [0.         0.         0.         0.        ]
 [0.30510

In [5]:
# test the agent with new q table

# Watch our agent play Frozen Lake by playing the best action 
# from each state according to the Q-table

for episode in range(3):
    # initialize new episode params
    state = env.reset()
    done = False
    print("*****EPISODE ", episode+1, "*****\n\n\n\n")
    time.sleep(1)

    for step in range(max_steps):    
        # Show current state of environment on screen
        # Choose action with highest Q-value for current state       
        # Take new action
        clear_output(wait=True)
        print(env.render())
        time.sleep(0.3)
        
        action=np.argmax(q_table[state,:])
        new_state,reward,done,info=env.step(action)

        if done:
            clear_output(wait=True)
            print(env.render())
            if reward == 1: 
                # Agent reached the goal and won episode
                print("****You reached the goal!****")
                time.sleep(3)
            else:
                # Agent stepped in a hole and lost episode            
                print("****You fell through a hole!****")
                time.sleep(3)
                clear_output(wait=True)
            break       

        # Set new state
        state = new_state

env.close()

  (Down)
SFFF
FHFH
FFFH
HFFG
None
****You reached the goal!****
